### PR TITLE
Warn when function returning python object is marked as noexcept

### LINKED ISSUE #5661: [BUG] Warn when noexcept is ignored
### Describe the bug

We should warn when `noexcept` is ignored - just to stop people thinking they're optimizing something that they aren't

### Code to reproduce the behaviour:

```cython
cdef object xx() noexcept:
    pass
```

This also applies to memoryviews

### Expected behaviour

This should generate a warning since exceptions are always checked for functions returning an object

### OS

Linux

### Python version

_No response_

### Cython version

current master

### Additional context

We might have to be a bit careful about fused functions - we probably don't want to generate warnings when *some* of the return types are valid for noexcept

### PR DESCRIPTION
**Pull request by matusvalo**
Fixes #5661

### CODE REVIEW

**Comment by matusvalo** [comment]
Hmm now I am reading the issue, I think memoryviews should be added too...

**Comment by matusvalo** [comment]
It seems that the tests are failing since following code snippet:
```cython
cdef class Bar:
    pass
```
is giving following warning:
```
warning: (tree fragment):11:34: noexcept clause is ignored for function returning Python object
```

**Comment by da-woods** [comment]
Most likely it's the auto-generated pickling functions. They're probably just implicitly noexcept (so I'd exclude anything that's implicitly noexcept from the warning)

**Comment by da-woods** [comment]
Probably https://github.com/cython/cython/blob/dd1b87a621e5139fd2cc48e67efbd2a754a202b5/Cython/Compiler/ParseTreeTransforms.py#L2236

**Comment by matusvalo** [comment]
> hey're probably just implicitly noexcept (so I'd exclude anything that's implicitly noexcept from the warning)

Right :-/. It is failng for correct scenario:

```cython
cdef test_ok():
    pass
```

This needs to be fixed...

**Comment by scoder** [comment]
>Most likely it's the auto-generated pickling functions. They're probably just implicitly noexcept

They shouldn't be, though. They can raise exceptions, and should be able to.

We should make sure that we're not covering up an actual bug here.

**Comment by da-woods** [comment]
> They shouldn't be, though. They can raise exceptions, and should be able to.

I think the better description is "it's always been obvious to Cython that a cdef function returning an object can raise an exception, so it's never had documented this by recording an exception specification". I think the functions can raise exceptions and there probably isn't a real bug. I'll double check this though

File: Cython/Compiler/Nodes.py [hunk 1]
```diff
@@ -625,6 +625,7 @@ class CFuncDeclaratorNode(CDeclaratorNode):
     # has_varargs      boolean
     # exception_value  ConstNode or NameNode    NameNode when the name of a c++ exception conversion function
     # exception_check  boolean or "+"    True if PyErr_Occurred check needed, "+" for a c++ check
+    # exc_clause       boolean    True if exception clause is explicitly declared
```
Comment (da-woods) [inline Cython/Compiler/Nodes.py:628]: I don't really like the name - maybe `has_explicit_exc_clause`?

I also have a vague memory we already have this information somewhere - I'll try to have a look later. I could be misremembering though.

**Review by da-woods — COMMENTED** [review]
I think it's probably also worth having a test with `legacy_implicit_noexcept` on - I don't think this should raise an error but I think it's worth testing.

File: Cython/Compiler/Nodes.py (hunk 1 above)
Comment (matusvalo) [reply Cython/Compiler/Nodes.py:628]: Parser seems to not pass further exc_clause:
https://github.com/cython/cython/blob/e9b08d3f071ebfad7f4886e43173cd0f72df6676/Cython/Compiler/Parsing.py#L2960-L2976

**Comment by matusvalo** [comment]
CI now is green.

> I think it's probably also worth having a test with legacy_implicit_noexcept on

I will have a look on that.

**Comment by matusvalo** [comment]
Tests for legacy_implicit_noexcept added. I think I am done here.

**Review by da-woods — COMMENTED** [review]
Looks good I think

**Comment by matusvalo** [comment]
Thank you @da-woods for your review. I will merge it and create backport PR to 3.0.X branch.

**Comment by tornaria** [comment]
This PR breaks sagemath https://github.com/sagemath/sage/issues/37560 in a way difficult to fix without switching off `legacy_implicit_noexcept` (and we are on RC stage so changing that is a no-go, but I'd still like to have a workaround for the release).

This is what I think is the root cause of the problem. Consider these two snippets:
```
cdef test_noexcept_warning():
    pass
```
and
```
cdef test_noexcept_warning() noexcept:
    pass
```
Which one is correct? The new warning tells me that the second one is incorrect, so I guess the first one is the correct one. However, when using `legacy_implicit_noexcept=True` the first one tells me `Function declaration should contain 'noexcept' keyword.`

In fact, this is how sagemath ended up with lots of functions as in the second example. Because we switched to cython 3 with `legacy_implicit_noexcept` and just added `noexcept` following the warning suggestion, on the belief that this was the correct version and preparing the way to disable `legacy_implicit_noexcept` in the future.

Would it make sense for `legacy_implicit_noexcept` to *not* emit the warning for this case? Then we can just follow the warnings to help us decide when to add `noexcept` and when to remove it in order to obtain a behavior that is equivalent to cython 0.29 as much as possible.

Looking to the future, a desirable property of the warnings would be: if there are no warnings when running with `legacy_implicit_noexcept=True`, then removing `legacy_implicit_noexcept` will not change the behavior.

----

Now as for why this is breaking sagemath. The problem is that this new warning cannot be disabled. We have a lot of doctests that (explicitly or implicitly) compile cython code, so these warnings break doctests, and I haven't found a way to disable this warning.

**Comment by scoder** [comment]
> I'd still like to have a workaround for the release

Do you need anything from Cython 3.0.9 or can you live with using 3.0.8 for the imminent sagemath release?

Regardless, removing the `noexcept` clause from a function returning an object is really the right thing to do and cannot break anything. There is *always* an exception check when returning objects, whether or not you declare them as `noexcept`. Thus the new warning.

**Comment by tornaria** [comment]
Thanks both for your quick answers. I figured out a way to disable the warning in sagemath, so I'll be able to workaround the issue just fine.

Still, if you can (without hurries) fix the warnings so that they lead to the correct choice regarding `noexcept`. That would be great.

I see removing the `noexcept` clause for those functions is the right way to fix this. As a matter of fact, I added most of those `noexcept` myself in https://github.com/sagemath/sage/pull/36507. It turned out we were getting ~ 59k warnings (7k different occurrences) and I assumed that every occurrence of the warning meant adding `noexcept` was correct. Building sagemath with cython 3.0.9 now gives ~ 40k of the new warning (4k different occurrences), so it seems about 2/3rds of the added `noexcept` are incorrect...


The new warning is indeed a good thing, as it will instruct me on which `noexcept` to remove, but I'd rather do that when removing it won't trigger a different warning.

The end goal is, eventually, to remove `legacy_implicit_noexcept`. The strategy I imagine is: fix everything until building gives no warnings. Then remove the option. Note that some of the warnings come from dependencies which we don't control (e.g. numpy and Cython itself).

I'm assuming that removing `legacy_implicit_noexcept` when there are still warnings is a bit risky.


> Do you need anything from Cython 3.0.9 or can you live with using 3.0.8 for the imminent sagemath release?

I don't have choice when using cython from the system. It turns out my distro (void linux) has already updated cython to 3.0.9. In general, we try that sagemath works fine with a range of versions of dependencies including the most recent release.

**Comment by tornaria** [comment]
What do you think of something like this:
```diff
diff --git a/Cython/Compiler/Nodes.py b/Cython/Compiler/Nodes.py
index 8cc824f6e..7ff0130d5 100644
--- a/Cython/Compiler/Nodes.py
+++ b/Cython/Compiler/Nodes.py
@@ -704,6 +704,19 @@ class CFuncDeclaratorNode(CDeclaratorNode):
 
         exc_val = None
         exc_check = 0
+
+        if (env.directives["legacy_implicit_noexcept"]
+                and not return_type.is_pyobject
+                and not self.has_explicit_exc_clause
+                and self.exception_check
+                and visibility != 'extern'):
+            # implicit noexcept, with a warning
+            self.exception_check = False
+            warning(self.pos,
+                    "Implicit noexcept declaration is deprecated."
+                    " Function declaration should contain 'noexcept' keyword.",
+                    level=2)
+
         if self.exception_check == '+':
             env.add_include_file('ios')         # for std::ios_base::failure
             env.add_include_file('new')         # for std::bad_alloc
diff --git a/Cython/Compiler/Parsing.py b/Cython/Compiler/Parsing.py
index 0711855a8..ca8b3b9ea 100644
--- a/Cython/Compiler/Parsing.py
+++ b/Cython/Compiler/Parsing.py
@@ -3141,9 +3141,7 @@ def p_exception_value_clause(s, is_extern):
                 exc_check = False
             # exc_val can be non-None even if exc_check is False, c.f. "except -1"
             exc_val = p_test(s)
-    if not is_extern and not exc_clause and s.context.legacy_implicit_noexcept:
-        exc_check = False
-        warning(s.position(), "Implicit noexcept declaration is deprecated. Function declaration should contain 'noexcept' keyword.", level=2)
+
     return exc_val, exc_check, exc_clause
 
 c_arg_list_terminators = cython.declare(frozenset, frozenset((
```

**Comment by tornaria** [comment]
> > What do you think of something like this: […]
> 
> Seems a reasonable chain of conditions to me. Could you send a PR with this change?

Will do later. I've been testing this for building sagemath and it seems to be working really good, as in: all the warnings now make sense, and it's possible to fix all of them and get no warnings at all.

**Pull request merged**